app/controllers/uphold_connections_controller.rb

```diff
@@ -0,0 +1,64 @@
+class UpholdConnectionsController < ApplicationController
+  UUID_LENGTH = 36
+
+  def login
+    ip_whitelist = Rails.application.secrets[:api_ip_whitelist]
+    render(status: 401) and return unless ip_whitelist.nil? || request.ip.in?(ip_whitelist)
+    uphold_card_id = params[:uphold_card_id]
+    render(status: 400, json: {}) and return unless uphold_card_id&.length == UUID_LENGTH
+    state = SecureRandom.hex(64).to_s
+    Rails.cache.write(state, uphold_card_id, expires_in: 10.minutes)
+    redirect_to Rails.application.secrets[:uphold_authorization_endpoint].
+      gsub('<UPHOLD_CLIENT_ID>', Rails.application.secrets[:uphold_login_client_id]).
+      gsub('<UPHOLD_SCOPE>', Rails.application.secrets[:uphold_scope]).
+      gsub('<STATE>', state)
+  end
+
+  def confirm
+    uphold_card_id = Rails.cache.fetch(params[:state])
+    uphold_connection, uphold_card = get_card(
+      uphold_card_id: uphold_card_id,
+      uphold_code: params[:code]
+    )
+
+    return redirect_to(root_path, flash: { alert: I18n.t(".out_of_time_alert") }) if uphold_card.blank?
```

Comment:
Should this be `redirect_to ... and return` based on the comment you made above?